astropy/table/table.py

```diff
@@ -3021,8 +3048,70 @@ def __eq__(self, other):
 
         return result
 
-    def __ne__(self, other):
-        return ~self.__eq__(other)
+    def cols_equal(self, other):
+        """
+        Element-wise comparison of table with another table, list, or scalar.
+
+        Returns a ``Table`` with the same columns containing boolean values
+        showing result of comparison.
+
+        Parameters
+        ----------
+        other : Table-like object or list or scalar
+             Object to compare with table
+
+        Examples
+        --------
+        Compare one Table with other::
+
+          >>> t1 = Table([[1, 2], [4, 5], [-7, 8]], names=('a', 'b', 'c'))
+          >>> t2 = Table([[1, 2], [-4, 5], [7, 8]], names=('a', 'b', 'c'))
+          >>> t1.cols_equal(t2)
+          <Table length=2>
+           a     b     c
+          bool  bool  bool
+          ---- ----- -----
+          True False False
+          True  True  True
+
+        """
+        # Check the length of other
+        check_len = True
+
+        if isinstance(other, Table):
+            names = other.colnames
+        else:
+            try:
+                other = Table(other, copy=False)
+                names = other.colnames
+            except Exception:
+                # Broadcast other into a dict, so e.g. other = 2 will turn into
+                # other = {'a': 2, 'b': 2} and then equality does a
+                # column-by-column broadcasting.
+                names = self.colnames
+                other = {name: other for name in names}
+                check_len = False
+
+        if check_len and len(self) != len(other):
+            raise ValueError('cannot compare tables with different lengths')
+
+        # Require column names match but do not require same column order
+        if set(self.colnames) != set(names):
+            raise ValueError('cannot compare tables with different column names')
+
+        eqs = []
+        for name in names:
+            try:
+                eq = self[name] == other[name]
+                assert isinstance(eq, np.ndarray) and eq.dtype is np.dtype('bool')
```

Comment:
This should be moved to after the `try/except` and raise a proper exception (or the result should be broadcast).  Note that numpy can give a deprecation warning and a boolean if things do not match. E.g., 
```
Column(np.arange(3.)) == np.array((1,2),'i4,i4')
/data/mhvk/venv/astropy-dev/lib/python3.7/site-packages/astropy-4.0.dev26169-py3.7-linux-x86_64.egg/astropy/table/column.py:988: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = getattr(super(), op)(other)
/usr/bin/ipython3:1: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
# False
```

Reply:
Broadcasting the result seems like a dodgy idea.  If the result is not an ndarray(bool) of the right length then basically the comparison failed for reasons that we cannot be sure of.  Numpy is really not helping out here...
```
In [25]: Column(np.arange(2)) == np.array([(1,2),(2,2)],'i4,i4')  # Should just return [False, False] ??
/Users/aldcroft/miniconda3/envs/astropy/bin/ipython:1: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
Out[25]: False

In [26]: Column(np.arange(3)) == np.array([(1,2),(2,2)],'i4,i4')  # Should raise a cannot broadcast exception
/Users/aldcroft/miniconda3/envs/astropy/bin/ipython:1: FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.
Out[26]: False

In [27]: Column(np.arange(3)) == np.array([(1,2)],'i4,i4')  # Should be [False, False, False]
# Same
False
```
So broadcasting the result would allow for comparing a length=3 table to [1,2] and it would happily return all False.  The current logic fails with "can't compare that".  An oddball case like comparing a length=2 table to `[(1,2), (3,4)], 'i4,i4'` is going to have problems in general because of the underlying numpy issues, but note that if the dtypes really do match then the comparison works as expected.

I would propose that for this release we just stick with this logic.  If you'd like I can change the assert to a test like:
```
if not (isinstance(eq, np.ndarray) and eq.dtype is np.dtype('bool')):
    raise TypeError(f'comparison for column {name} returned {eq} instead of the expected boolean ndarray')
```

Reply:
I think with your idea of in future adding a join_type like argument, we can indeed for now stick to just raising an exception.